File: byteback-test/test-java-8/src/main/java/byteback/test/exceptions/PotentialUnsafeCast.java
/**
 * RUN: %{byteback} -cp %{jar} -c %{class} --cce -o %t.bpl
 */
package byteback.test.exceptions;

import byteback.specification.Contract.Behavior;
import byteback.specification.Contract.Raise;
import byteback.specification.Contract.Raises;
import byteback.specification.Contract.Return;
import static byteback.specification.Operators.*;

@SuppressWarnings("unused")
public class PotentialUnsafeCast {

	public static class A {

		@Return
		public A() {}

	}

	public static class B extends A {

		@Return
		public B() {}

	}

	public static class C {

		@Return
		public C() {}

	}

	@Behavior
	public boolean x_is_instanceof_A(final Object x) {
		return x instanceof A;
	}

	@Behavior
	public boolean x_is_not_instanceof_A(final Object x) {
		return not(x_is_instanceof_A(x));
	}

	@Return(when = "x_is_instanceof_A")
	public void unsafeDownCast(final Object x) {
		A a = (A) x;
	}

	@Return
	public void safeDownCast(final Object x) {
		if (x instanceof A) {
			A a = (A) x;
		}
	}

	@Return
	public void safeUpcast(final B b) {
		A a = (A) b;
	}

	@Behavior
	public boolean always() {
		return true;
	}

	@Raise(exception = ClassCastException.class, when = "always")
	public void invalidDownCast() {
		A a = new A();
		B b = (B) a;
	}

	@Return
	public void validCastNull() {
		A a = (A) null;
	}

}
/**
 * RUN: %{verify} %t.bpl | filecheck %s
 * CHECK: Boogie program verifier finished with 6 verified, 0 errors
 */

File: byteback-test/test-kotlin-1.8.0/src/main/kotlin/byteback/test/exceptions/PotentialUnsafeCast.kt
/**
 * RUN: %{byteback} -cp %{jar} -c %{class} -c %{class}$A -c %{class}$B --cce -o %t.bpl
 */
package byteback.test.exceptions;

import byteback.specification.Contract.*;
import byteback.specification.Operators.*;

@SuppressWarnings("unused")
class PotentialUnsafeCast {

		open class A {
				@Return
				constructor()
		}

		class B : A() {
		}

		class C {
		}

		@Behavior
		fun `x is instanceof A`(x: Any): Boolean {
				return x is A
		}

		@Return(`when` = "x is instanceof A")
		fun unsafeDownCast(x: Any): Unit {
				var a: A = x as A
		}

		@Return
		fun safeDownCast(x: Any): Unit {
				if (x is A) {
						var a: A = x as A
				}
		}

		@Return
		fun safeUpcast(b: B): Unit {
				var a: A = b as A
		}

		@Behavior
		fun always(): Boolean {
				return true
		}

		@Raise(`exception` = ClassCastException::class, `when` = "always")
		fun invalidDownCast(): Unit {
				var a: A = A()
				var b: B = a as B
		}

}
/**
 * RUN: %{verify} %t.bpl | filecheck %s
 * CHECK: Boogie program verifier finished with 7 verified, 0 errors
 */
